00_homework_kawapa/Z4.cpp

```diff
@@ -0,0 +1,84 @@
+#include <chrono>
+#include <iostream>
+#include <sstream>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+constexpr size_t n = 15;
+// number of philosophers and forks
+
+std::mutex forks[n];
+
+struct Philosopher
+{
+    int id_;
+    int forkIndexLeft_;
+    int forkIndexRight_;
+    bool hungry_ = true;
+
+    void dine()
+    {
+        while (hungry_)
+        {
+            contemplate();
+            eat();
+        }
+    }
+
+    void eat()
+    {
+        wait();
+        std::lock(forks[forkIndexLeft_], forks[forkIndexRight_]);
+ 
+        std::lock_guard<std::mutex> mutex_l(forks[forkIndexLeft_], std::adopt_lock);
+        print("has just picked up the left fork");
+
+        wait();
+        std::lock_guard<std::mutex> mutex_r(forks[forkIndexRight_], std::adopt_lock);
+        print("has just picked up the right fork and he starts eating");
+
```

Comment:
trochę sztuczne te opóźnienia z wait().  Tak naprawdę to po zakończeniu funkcji lock już mamy oba mutexy. lock_guardy są tylko, żeby RAII zapewnić, ale gdyby print lub wiat rzuciło wyjątkiem to na nic by się to zabezpieczenie nie przydało. Użyj najlepiej std::scoped_lock